File: frontend/src/types/reports.ts
export type ReportPeriod = 'day' | 'week' | 'month';
export type ReportBucketSize = 'day' | 'week';

export interface RevenuePoint {
  period_start: string;       // ISO date — first day of the bucket
  revenue_cents: number;
  payments_count: number;
}

export interface PlanPopularity {
  plan_id: number;
  plan_name: string;
  active_members: number;
  revenue_cents: number;
}

export interface RetentionSummary {
  active: number;
  cancelled: number;
  expired: number;
  pending: number;
  total_lifetime: number;
  churn_rate: number;
}

export interface ReportsResponse {
  period: ReportPeriod;
  bucket_size: ReportBucketSize;
  currency: string;
  revenue_by_period: RevenuePoint[];
  popular_plans: PlanPopularity[];
  retention: RetentionSummary;
}

File: backend/app/schemas/reports.py
"""Schemas for the admin reports endpoint (Phase 9)."""

from __future__ import annotations

from pydantic import BaseModel


class RevenuePoint(BaseModel):
    period_start: str  # ISO date — first day of the bucket
    revenue_cents: int
    payments_count: int


class PlanPopularity(BaseModel):
    plan_id: int
    plan_name: str
    active_members: int
    revenue_cents: int


class RetentionSummary(BaseModel):
    active: int
    cancelled: int
    expired: int
    pending: int
    total_lifetime: int
    churn_rate: float  # cancelled / total_lifetime, 0..1


class ReportsResponse(BaseModel):
    """The full reports payload for the requested window."""

    period: str            # day | week | month (the requested range)
    bucket_size: str       # day | week (granularity)
    currency: str
    revenue_by_period: list[RevenuePoint]
    popular_plans: list[PlanPopularity]
    retention: RetentionSummary
